packages/xstate-react/src/useInterpret.ts

```diff
@@ -119,6 +125,12 @@ export function useInterpret<
     };
   }, []);
 
+  useIsomorphicLayoutEffect(() => {
+    if (input) {
+      service.input(input);
+    }
+  }, [...Object.keys(input || {}), ...Object.values(input || {})]);
```

Comment:
As part of the "dont allow users to write weird code"... I would just check the values here. The "input" object should really be only used inline and should always contain the same keys. Even if we assume that this wouldn't be a requirement then this code would still not be entirely correct - if we allow the shape of the `input` to change then we can't rely on spreading those things here because we can't be sure about the number of keys/values (it could change between renders) and that breaks the "rules of hooks".

So in a sense - we already make an assumption that the input object is always of the same shape, so let's just embrace it more

Reply:
Yes, makes sense.
```suggestion
  }, [...Object.values(input || {})]);
```